demos/common/python/models/hpe_hrnet.py

```diff
@@ -0,0 +1,136 @@
+
+        self.decoder = AssociativeEmbeddingDecoder(
+            num_joints=self.net.outputs[self.heatmaps_blob_name].shape[1],
+            adjust=True,
+            refine=True,
+            delta=0.5,
+            max_num_people=30,
+            detection_threshold=0.1,
+            tag_threshold=1,
+            pose_threshold=prob_threshold,
+            use_detection_val=True,
+            ignore_too_much=False,
+            dist_reweight=True)
+        self.size_divisor = size_divisor
+
+    @staticmethod
+    def _get_inputs(net):
+        image_blob_name = None
+        for blob_name, blob in net.input_info.items():
+            if len(blob.input_data.shape) == 4:
+                image_blob_name = blob_name
+            else:
+                raise RuntimeError('Unsupported {}D input layer "{}". Only 4D input layers are supported'
+                                   .format(len(blob.shape), blob_name))
+        if image_blob_name is None:
+            raise RuntimeError('Failed to identify the input for the image.')
+        return image_blob_name
+
+    def preprocess(self, inputs):
+        img = resize_image(inputs, (self.w, self.h), keep_aspect_ratio=True)
+        h, w = img.shape[:2]
+        if not (self.h - self.size_divisor < h <= self.h and self.w - self.size_divisor < w <= self.w):
+            self.logger.warn("Chosen model aspect ratio doesn't match image aspect ratio")
+        resize_img_scale = np.array((inputs.shape[1] / w, inputs.shape[0] / h), np.float32)
+
+        pad = [(self.h - h + 1) // 2, (self.w - w + 1) // 2, (self.h - h) // 2, (self.w - w) // 2]
+        img = np.pad(img, ((pad[0], pad[2]), (pad[1], pad[3]), (0, 0)),
+                     mode='constant', constant_values=0)
+        img = img.transpose((2, 0, 1))  # Change data layout from HWC to CHW
+        img = img[None]
+        meta = {
+            'original_size': inputs.shape[:2],
+            'resize_img_scale': resize_img_scale
+        }
+        return {self.image_blob_name: img}, meta
+
+    def postprocess(self, outputs, meta):
+        heatmaps = outputs[self.heatmaps_blob_name]
+        nms_heatmaps = outputs[self.nms_heatmaps_blob_name]
+        aembds = outputs[self.embeddings_blob_name]
+
+        poses, scores = self.decoder(heatmaps, aembds, nms_heatmaps=nms_heatmaps)
+        # Rescale poses to the original image.
+        poses = self.transform_preds(poses, meta)
+        return poses, scores
+
+    def transform_preds(self, poses, meta):
+        # compute trans
+        scale = min(meta['resize_img_scale'])
+        trans = np.zeros((2, 3))
+        trans[0][0] = trans[1][1] = scale * self.output_scale
+
+        shift = (max(meta['original_size']) - max(self.h, self.w) * scale) / 2
+        trans[np.argmax(meta['original_size'][::-1])][2] = shift
+
+        # affine_trans
+        for person in poses:
+            for joint in person:
+                joint[:2] = self.affine_transform(joint[:2], trans)
```

Comment:
To me affine transformations are always scary :) Direct addition and multiplication should be easier to understand. Plus it will align with `HpeAssociativeEmbedding` even more because there will only be difference in shift.